Let .kubeconfig be a single-source config for API clients

liggitt opened the pull request:
`client.Config` supports setting the actual data for CA, client cert and key as an alternative to using file references. This updates .kubeconfig to be able to do the same. This is primarily useful for encapsulating a complete client config in a single file (e.g. for distribution to a node, pod, or other API client).

I intentionally didn't update `kubectl config` commands to allow setting inline cert data since I expect this to be used by something that generates these files, rather than end-users. If they really really want to set the data, they can use `kubectl config set/unset`.

Finally, I added support for basic-auth username/password in .kubeconfig (previously the only way to set those was to reference a second .kubernetes_auth file).

liggitt committed:
Let .kubeconfig populate ca/cert/key data, and basic-auth username/password in client configs

liggitt commented:
@googlebot I'm with RedHat
/cc @derekwaynecarr now you've got me talking to the bot

liggitt commented:
/cc @deads2k

nikhiljindal commented:
Not sure who is a good person to assign this review to.
@erictune?

deads2k commented:
@nikhiljindal 

@jlowdermilk has been the victim in this area several times.

deads2k commented on pkg/client/clientcmd/validation.go:
```diff
@@ -109,7 +109,8 @@ func validateClusterInfo(clusterName string, clusterInfo clientcmdapi.Cluster) [
 		validationErrors = append(validationErrors, fmt.Errorf("no server found for %v", clusterName))
 	}
-	if len(clusterInfo.CertificateAuthority) != 0 {
+	// Only validate the CA file if it hasn't been overridden by CA data
```
I'd rather see an either or.  I don't think it makes much sense to allow both a CA file and CA data to be specified.

liggitt replied:
Where would you enforce that? I could see the either/or when building the `client.Config`

deads2k replied:
I don't see a reason why a .kubeconfig file would contain both certfile and certdata for a single user stanza.

liggitt replied:
I agree it doesn't make sense. Where would you enforce it?

deads2k replied:
Is there a reason not to do it in the validateAuthInfo method of the validation.go file that is use to validate that a .kubeconfig is properly formed, right where we're commenting?

liggitt replied:
looks like validation only runs when it's turning a .kubeconfig into a client.Config, not when using `kubectl config` commands to update the config. I updated the validation and updated the `kubectl` commands to clear the *data fields when setting the file fields.

smarterclayton commented:
This is on the critical path to giving pods access to talk back to the master (as in, now .kubeconfig is a nice self contained bundle that we can ship around as a secret for @pmorie's secret change)

liggitt commented:
@deads2k updated and updated tests:

* `set-cluster` disallows `--insecure=true` and `--certificate-authority=...`
* `set-cluster --insecure=true` clears CA data and file
* `set-cluster --certificate-authority=...` sets insecure=false and clears CA data
* `set-credentials` disallows `--token=...` and `--client-certificate=...`
* `set-credentials --token=...` clears client certificate and key files and data
* `set-credentials --client-certificate=...` clears token

derekwaynecarr commented:
Jordan - per our chat today, can we allow a simple user name and password for clusters using basic auth over SSL as part of this PR?

Sent from my iPhone

> On Feb 18, 2015, at 2:55 PM, Jordan Liggitt <notifications@github.com> wrote:
> 
> @deads2k updated and updated tests:
> 
> set-cluster disallows --insecure=true and --certificate-authority=...
> set-cluster --insecure=true clears CA data and file
> set-cluster --certificate-authority=... sets insecure=false and clears CA data
> set-credentials disallows --token=... and --client-certificate=...
> set-credentials --token=... clears client certificate and key files and data
> set-credentials --client-certificate=... clears token
> —
> Reply to this email directly or view it on GitHub.
>

liggitt commented:
@derekwaynecarr added

liggitt commented:
@deads2k I was hoping basic auth support would go away with .kubernetes_auth, but enough systems use it that that seems unlikely. Adding first-class support in .kubeconfig hopefully means we can deprecate .kubernetes_auth more easily.

derekwaynecarr commented:
Awesome to know it's already there.  

Can we get an UPSTREAM patch in Origin tomorrow?

Sent from my iPhone

> On Feb 18, 2015, at 10:04 PM, Jordan Liggitt <notifications@github.com> wrote:
> 
> @derekwaynecarr added
> 
> —
> Reply to this email directly or view it on GitHub.
>

jlowdermilk commented on pkg/kubectl/cmd/config/create_authinfo.go:
```diff
@@ -35,20 +36,34 @@ type createAuthInfoOptions struct {
 
 	cmd := &cobra.Command{
-		Use:   fmt.Sprintf("set-credentials name [--%v=path/to/auth/file] [--%v=path/to/certficate/file] [--%v=path/to/key/file] [--%v=bearer_token_string]", clientcmd.FlagAuthPath, clientcmd.FlagCertFile, clientcmd.FlagKeyFile, clientcmd.FlagBearerToken),
+		Use:   fmt.Sprintf("set-credentials name [--%v=path/to/auth/file] [--%v=certfile] [--%v=keyfile] [--%v=bearer_token] [--%v=basic_user] [--%v=basic_password]", clientcmd.FlagAuthPath, clientcmd.FlagCertFile, clientcmd.FlagKeyFile, clientcmd.FlagBearerToken, clientcmd.FlagUsername, clientcmd.FlagPassword),
```
path/to/auth/file -> authfile to match the other flags.

jlowdermilk commented on pkg/kubectl/cmd/config/create_authinfo.go:
```diff
@@ -95,17 +112,55 @@ func (o createAuthInfoOptions) run() error {
+		setBasic = setBasic || len(modifiedAuthInfo.Password) > 0
+	}
+
+	// If any auth info was set, make sure any other existing auth types are cleared
```
Would be better to throw an error if mutually exclusive auth types were provided. If the user's input is invalid, it's less surprising to fail early with an actionable message than silently ignore values.

jlowdermilk replied:
Nevermind, l see validate() does this.

jlowdermilk commented on pkg/kubectl/cmd/config/create_authinfo.go:
```diff
@@ -35,20 +36,34 @@ type createAuthInfoOptions struct {
+
+    set-credentials cluster-admin --client-key=~/.kube/admin.key
+
+  Only one type of credentials can be set per user:
```
Bearer token and basic auth are mutually exclusive, but is client cert supposed to be mutually exclusive with them as well? The current auth scheme used by kubectl includes both a client cert and basic auth; I'm pretty sure it fails if either is missing.

liggitt replied:
@jlowdermilk good catch... looking at how the transport is built, cert is orthogonal to bearer/basic. will update

liggitt commented:
@jlowdermilk updated validation and `kubectl set-credentials` to allow both client cert info and basic/bearer auth. let me know if you have any other comments

smarterclayton commented:
This is LGTM, but @jlowdermilk owns final sign off and review

liggitt commented:
A couple final tweaks regarding loading/writing kubeconfig files
1. Changed the default mode when creating a kubeconfig file 0600... world-readable files aren't great, especially when they can contain secrets (which it already did with `bearer-token`)
2. Added methods for loading/writing a config to a byte slice, for consumers that want to serialize/deserialize from things other than files

jlowdermilk commented:
LGTM.
